---
title: 'Practical 03 SG: LD and Haplotype estimation'
author: "Kathryn Weissman & Irene Fernández"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list=ls())
library(genetics)
library(HardyWeinberg)
library(data.table)
library(LDheatmap)
library(knitr)
library(haplo.stats)
```


# LD (15p.)

1. The file `FOXP2.zip` contains genetic information of individuals of a Japanese population of unrelated
individuals. The genotype information concerns SNPs of the Forkhead box protein P2 (FOXP2) gene
region, located the long arm of chromosome number 7. This gene plays an important role in the
development of speech and language. The `FOXP2.zip` file contains:

* `FOXP2.dat`: a text file with the genotype data which can be read in with R.
* `FOXP2.fam`: a PLINK file with data on the individuals (family id, individual id, ids of parents, sex and
phenotype).
* `FOXP2.bed`: a PLINK file with binary genotype data.
* `FOXP2.bim`: a PLINK file with data on the genetic variants (chromosome, SNP identifier, basepair
position along the chromosome and alleles).


2. (1p) Load the **FOXP2.dat** file into the R environment. How many individuals and how many SNPs
are there in the database? What percentage of the data is missing?

* There are 104 individuals in the database.
* There are 543 SNPs in the database.
* 0% of the data is missing.

```{r}
file <- "~/Downloads/FOXP2.dat"
data <- fread(file, header = TRUE, index="id")
data[1:10, 1:10]
```

```{r}
n <- nrow(data); n # number of samples
p <- ncol(data) - 1; p # number of SNPs - first column is id
perc.mis <- 100*sum(is.na(data))/(n*p); perc.mis # percentage of missing data overall
```

3. (1p) Determine the genotype counts for each SNP, and depict all SNPs simultaneously in a ternary
plot, and comment on your result. For how many variants do you reject Hardy-Weinberg equilibrium
using an ordinary chi-square test without continuity correction? (hint: you can read the `.bim` in R
in order to determine the alleles of each SNP, and use function `MakeCounts` from the `HardyWeinberg`
package to create a matrix of genotype counts).

* The green dots represent SNPs that are in equilibrium, and the red dots represent SNPs that are out
of equilibrium.
in the data set where the B allele frequency is greater than 0.5.
* Using an ordinary chi-square test we reject Hardy-Weinberg equilibrium for 33 variants.
* *Note:* In order to create the genotype counts matrix, we create a function similar to the one of the
anterior practical.

```{r}
genotype_matrix <- matrix(ncol = 3, nrow = 0, dimnames = list(NULL, c("AA", "AB", "BB")))
for(i in colnames(data[,-1])){
snp <- genotype(data[[i]], reorder = "ascii") # reorder is to use as specific order for the alleles
snp.counts <- c(sum(snp==paste(allele.names(snp)[1], allele.names(snp)[1], sep = "/")),
                sum(snp==paste(allele.names(snp)[1], allele.names(snp)[2], sep = "/")),
                sum(snp==paste(allele.names(snp)[2], allele.names(snp)[2], sep = "/")))
genotype_matrix <- rbind(genotype_matrix, snp.counts)
}
rownames(genotype_matrix) <- colnames(data[,-1])
```

```{r}
ternaryPlot <- HWTernaryPlot(genotype_matrix)
```

```{r}
results_ChiSq <- HWChisqStats(genotype_matrix, pvalues=TRUE)
sum(results_ChiSq <= .05)
```

4. (1p) Using the function `LD` from the `genetics` package, compute the LD statistic *D* for the SNPs
rs34684677 and rs2894715 of the database. Is there significant association between the alleles of these
two SNPs?

Yes, there is association between the alleles because the p-value is small and D' is close to 1.

```{r}
LD(genotype(data$rs34684677), genotype(data$rs2894715))
```

5. (2p) Also compute the LD statistic *D* for the SNPs rs34684677 and rs998302 of the database. Is there
significant association between these two SNPs? Is there any reason why rs998302 could have stronger
or weaker correlation than rs2894715?

These two SNPS are not significantly associated because the p-value is large. rs34684677 and rs998302 may have a weaker correlation because the sites are located further from each other or they don't share any functional relationship.

```{r}
LD(genotype(data$rs34684677), genotype(data$rs998302))
```

6. (2p) Given your previous estimate of *D* for SNPs rs34684677 and rs2894715, infer the haplotype
frequencies. Which haplotype is the most common?

```{r}
SNP1 <- genotype(data$rs34684677)
SNP2 <- genotype(data$rs2894715)
summary(SNP1)
summary(SNP2)
LD_res <- LD(genotype(data$rs34684677), genotype(data$rs2894715))

pA <- summary(SNP1)$allele.freq[,2][1]
pB <- summary(SNP2)$allele.freq[,2][1]
pa <- summary(SNP1)$allele.freq[,2][2]
pb <- summary(SNP2)$allele.freq[,2][2]

pAB <- LD_res$D + pA*pB #D = pAB - pApB
pab <- LD_res$D + pa*pb #D = pab - papb
```
* The haplotype GT have a frequency of `r pAB`.
* The haplotype TG have a frequency of `r pab`.
* As we could observe, the most common haplotype is the GT.

7. (2p) Compute the LD statistics $R^2$ for all the marker pairs in this data base, using the `LD` function of
the packages `genetics`. Be prepared that this make take a few minutes. Also compute an alternative
estimate of $R^2$ obtained by using the `PLINK` program. For this purpose you should:

* Download and install `PLINK` 1.90 from https://www.cog-genomics.org/plink2/
* Take care to store the files `FOXP2.bim`, `FOXP2.fam` and `FOXP2.bed` in a directory where `PLINK` can
find them.
* Compute LD estimates with `PLINK` using `plink --bfile FOXP2 --r2 --matrix --out FOXP2` This
creates a file with extension `FOXP2.ld` that contains a matrix with all $R^2$ statistics. Read this file into
the R environment. Make a scatter plot for R’s LD estimates against PLINK’s LD estimates. Are they
identical or do they at least correlate? What’s the difference between these two estimators? Which
estimator would your prefer and why?

* As the plot shows, the $R^2$ statistics are very similar with a correlation of 0.99. If the results do not differ, we prefer PLINK because is faster in the calculation.

```{r}
# Convert one SNP at a time to genotype and save it as a column in dataframe.
# RES <- data.frame(genotype(data[[2]],sep="/"))
# for(i in 3:ncol(data)) {
#    snp <- genotype(data[[i]],sep="/")
#    RES <- cbind(RES,snp)
# }
```

```{r}
#output <- LD(RES)
```

```{r}
#R2 <- output$"R^2"
#write.table(R2, file="FOXP2_R2_r.txt", row.names=TRUE, col.names=TRUE)
```

```{r}
R2 <- read.table("FOXP2_R2_r.txt", row.names = NULL)
#head(R2)
R2 <- R2[, -1]
dim(R2)
R2_plink <- read.table("FOXP2.ld")
#head(R2_plink)
dim(R2_plink)
for(i in 1:ncol(R2_plink)) {
  for(j in 1:i) {
    R2_plink[i,j] = NA
  }
}

plot(R2[!is.na(R2)], R2_plink[!is.na(R2_plink)])
cor(R2[!is.na(R2)], R2_plink[!is.na(R2_plink)])
```

8. (2p) Compute a distance matrix with the distance in base pairs between all possible pairs of SNPs,
using the basepair position of each SNP given in the `.bim` file. Make a plot of R’s $R^2$ statistics against
the distance (expressed as the number of basepairs) between the markers. Comment on your results.

```{r}
dist_data <- fread("~/Downloads/FOXP2.bim")
head(dist_data)
dist_data <- dist_data[match(colnames(data)[-1], dist_data$V2)]
dist_data <- dist_data[, 4]
x <- dist(dist_data)
head(x)
length(x)

plot(x, t(R2[!is.na(R2)]))
```

9. (2p) Make an LD heatmap of the markers in this database, using the $R^2$ statistic with the `LD` function.
Make another heatmap obtained by filtering out all variants with a MAF below 0.35, and redoing the
computations to obtain the $R^2$ statistics in R. Can you explain any differences observed between the
two heatmaps?

* In the first LDheatmap there is more uncorrelation in general, this could be because we are keeping those SNPs with lower maf and these SNPs not provide much variation to have a correlation with other SNPs.

```{r}
RES <- data.frame(genotype(data[[2]],sep="/"))
for(i in 3:ncol(data)) {
   snp <- genotype(data[[i]],sep="/")
   RES <- cbind(RES,snp)
}
rgb.palette <- colorRampPalette(rev(c("blue", "orange", "red")), space = "rgb")
#LDheatmap(gdat=RES,color=rgb.palette(18), LDmeasure = 'r')
include_graphics("LDheatmap.png", dpi = 200)

maf <- function(x){
  x <- genotype(x,sep="/")
  out <- summary(x)
  af1 <- min(out$allele.freq[,2],na.rm=TRUE)
  af1[af1==1] <- 0 
  return(af1)
}
mafs <- apply(data[,-1], 2, maf)
sub_snp <- mafs[mafs > 0.35]

data_maf <- as.data.frame(data)
data_maf <- data_maf[, names(sub_snp)]
RES_maf <- data.frame(genotype(data_maf[[1]],sep="/"))
for(i in 2:ncol(data_maf)) {
   snp <- genotype(data_maf[[i]],sep="/")
   RES_maf <- cbind(RES_maf,snp)
}
rgb.palette <- colorRampPalette(rev(c("blue", "orange", "red")), space = "rgb")
#LDheatmap(gdat=RES_maf,color=rgb.palette(18))
include_graphics("LDheatmap_maf.png", dpi = 200)
```

10. (1p) Can you distinguish blocks of correlated markers in the area of the FOXP2 gene? How many blocks
do you think that *at least* seem to exist?

* In the second plot we could distinguish 3 main blocks with high correlation, two of them also being correlated but in a weaker manner.

11. (1p) Simulate independent SNPs under the assumption of Hardy-Weinberg equilibrium, using R’s `sample`
instruction `sample(c("AA","AB","BB"),n,replace=TRUE,prob=c(p*p,2*p*q,q*q)))`. Simulate as
many SNPs as you have in your database, and take care to match each SNP in your database with
a simulated SNP that has the same sample size and allele frequency. Make an LD heatmap of the
simulated SNPs, using $R^2$ as your statistic. Compare the results with the LD heatmap of the FOXP2
region. What do you observe? State your conclusions.

* 

```{r}
f_sample_data <- function(x) {
  x = genotype(x, sep = "/")
  out = summary(x)
  p = out$allele.freq[1,2]
  q = out$allele.freq[2,2]
  A = out$allele.names[1]
  B = out$allele.names[2]
  AA = paste(A, A, sep = "/")
  AB = paste(A, B, sep = "/")
  BB = paste(B, B, sep = "/")
  return(sample(c(AA,AB,BB), nrow(data), replace=TRUE, prob=c(p*p,2*p*q,q*q)))
}
sample_data = data.frame(apply(as.data.frame(data[,-1]), 2, f_sample_data))
#head(sample_data)

RES_sample <- data.frame(genotype(sample_data[,1],sep="/"))
for(i in 2:ncol(sample_data)) {
  snp <- genotype(sample_data[,i],sep="/")
  RES_sample <- cbind(RES_sample,snp)
}
#LDheatmap(gdat=RES_sample,color=rgb.palette(18), LDmeasure = 'r')
include_graphics("LDheatmap_sample.png", dpi = 200)
```



# Haplotype estimation (10p.)

1. Apolipoprotein E (APOE) is a protein involved in Alzheimer’s disease. The corresponding gene *APOE*
has been mapped to chromosome 19. The file `APOE.dat` contains genotype information of unrelated
individuals for a set of SNPs in this gene. Load this data into the R environment. `APOE.zip` contains
the corresponding `.bim`, `.fam` and `.bed` files. You can use the `.bim` file to obtain information about
the alleles of each polymorphism.

```{r}
rm(list=ls())
file <- "~/Downloads/APOE.dat"
data <- as.data.frame(fread(file, header = TRUE))
data[1:10, 1:10]

data_id <- data$id
data <- data[,-1]
```

2. (1p) How many individuals and how many SNPs are there in the database? What percentage of the
data is missing?

* There are 107 individuals in the database.
* There are 161 SNPs in the database.
* 0% of the data is missing.

```{r}
n <- nrow(data); n # number of samples
p <- ncol(data); p # number of SNPs - first column is id
perc.mis <- 100*sum(is.na(data))/(n*p); perc.mis # percentage of missing data overall
```

3. (1p) Assuming all SNPs are bi-allelic, how many haplotypes can theoretically be found for this data set?

* Theoretically 5.846007e+48 haplotypes can be found for this dataset.
```{r}
2^p
```

4. (2p) Estimate haplotype frequencies using the `haplo.stats` package (set the minimum posterior probability
to 0.001). How many haplotypes do you find? List the estimated probabilities in decreasing order.
Which haplotype number is the most common?

* We find 31 haplotypes. 
* The most common haplotype is the number 27.

```{r}
Geno <- cbind(substr(data[,1],1,1),substr(data[,1],3,3))
for(i in 2:ncol(data)) {
  Geno <- cbind(Geno,substr(data[,i],1,1),substr(data[,i],3,3))
}
Haplo.Res <- haplo.em(Geno,locus.label=colnames(data),control=haplo.em.control(min.posterior=0.001))
#Haplo.Res
Haplo.Res$nreps
sort(Haplo.Res$hap.prob, decreasing = TRUE)
which(Haplo.Res$hap.prob == max(Haplo.Res$hap.prob))
```

5. (2p) Is the haplotypic constitution of any of the individuals in the database ambiguous or uncertain?
For how many? What is the most likely haplotypic constitution of individual NA20763? (identify the
constitution by the corresponding haplotype numbers).

* There are 19 ambiguous constitutions.
* The most likely haplotypic constitution of individual NA20763 is 8.

```{r}
sum(Haplo.Res$nreps > 1)
Haplo.Res$hap1code[which(data_id == "NA20763")]
```

6. (1p) Suppose we would delete polymorphism rs374311741 from the database prior to haplotype estimation.
Would this affect the results obtained? Justify your answer.

* Deleting one polymorphism generates very minimal variations (var = 0.005) in the estimated probabilities and there is still 31 haplotypes.

```{r}
data_sub <- data[, !(colnames(data) %in% c("rs374311741"))]
Geno_sub <- cbind(substr(data_sub[,1],1,1),substr(data_sub[,1],3,3))
for(i in 2:ncol(data_sub)) {
  Geno_sub <- cbind(Geno_sub,substr(data_sub[,i],1,1),substr(data_sub[,i],3,3))
}
Haplo.Res_sub <- haplo.em(Geno_sub,locus.label=colnames(data_sub),control=haplo.em.control(min.posterior=0.001))
#Haplo.Res_sub
Haplo.Res_sub$nreps
var(Haplo.Res_sub$hap.prob, Haplo.Res$hap.prob)
```

7. (1p) Remove all genetic variants that have a minor allele frequency below 0.10 from the database, and
re-run `haplo.em`. How does this affect the number of haplotypes?

* Removing the variants with MAF < 0.10, we only remain with 21 variants. And the number of haplotypes decreases to 8.

```{r}
maf <- function(x){
  x <- genotype(x,sep="/")
  out <- summary(x)
  af1 <- min(out$allele.freq[,2],na.rm=TRUE)
  af1[af1==1] <- 0
  af1
}
mafs <- apply(data, 2, maf)
data_maf <- data[, mafs >= 0.10]
dim(data_maf)

Geno_maf <- cbind(substr(data_maf[,1],1,1),substr(data_maf[,1],3,3))
for(i in 2:ncol(data_maf)) {
  Geno_maf <- cbind(Geno_maf,substr(data_maf[,i],1,1),substr(data_maf[,i],3,3))
}

Haplo.Res_maf <- haplo.em(Geno_maf,locus.label=colnames(data_maf),control=haplo.em.control(min.posterior=0.001))
#Haplo.Res_maf
Haplo.Res_maf$nreps
Haplo.Res_maf$hap.prob
```

8. (2p) We could consider the newly created haplotypes in our last run of `haplo.em` as the alleles of a new
superlocus. Which is, under the assumption of Hardy-Weinberg equilibrium, the most likely genotype
at this new locus? What is the probability of this genotype? Which genotype is the second most likely,
and what is its probability?

* The most likely genotype is the one of the haplotype 36 with probability of 0.38.
* The second most likely genotype is the one of the haplotype 8 with probability of 0.16.

```{r}
Haplo.Res_maf$haplotype
haplotypes <- Haplo.Res_maf$haplotype
probabilities <- Haplo.Res_maf$hap.prob

haplotype_pairs <- get.hapPair(haplotypes, probabilities, base.index=1)

genotype_prob <- haplotype_pairs$p.g
genotype1 <- which(genotype_prob == max(genotype_prob))
genotype2 <- which(genotype_prob == max(genotype_prob[-genotype1]))

haplotype_pairs$x.haplo[genotype1,]
genotype_prob[genotype1]
haplotype_pairs$x.haplo[genotype2,]
genotype_prob[genotype2]
```
